community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/NativeLabelScanReader.java

```diff
@@ -0,0 +1,165 @@
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+    }
+
+    @Override
+    public PrimitiveLongIterator nodesWithLabel( int labelId )
+    {
+        RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor;
+        try
+        {
+            ensureOpenCursorsClosed();
+            cursor = seekerForLabel( labelId );
+            openCursors.offer( cursor );
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+
+        return new LabelScanValueIterator( rangeSize, cursor );
+    }
+
+    @Override
+    public PrimitiveLongIterator nodesWithAnyOfLabels( int... labelIds )
+    {
+        List<PrimitiveLongIterator> iterators = iteratorsForLabels( labelIds );
+        return new CompositeLabelScanValueIterator( iterators, false );
+    }
+
+    @Override
+    public PrimitiveLongIterator nodesWithAllLabels( int... labelIds )
+    {
+        List<PrimitiveLongIterator> iterators = iteratorsForLabels( labelIds );
+        return new CompositeLabelScanValueIterator( iterators, true );
+    }
+
+    private List<PrimitiveLongIterator> iteratorsForLabels( int[] labelIds )
+    {
+        List<PrimitiveLongIterator> iterators = new ArrayList<>();
+        try
+        {
+            ensureOpenCursorsClosed();
+            for ( int labelId : labelIds )
+            {
+                RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor = seekerForLabel( labelId );
+                openCursors.offer( cursor );
+                iterators.add( new LabelScanValueIterator( rangeSize, cursor ) );
+            }
+        }
+        catch ( IOException e )
+        {
+            throw new RuntimeException( e );
+        }
+        return iterators;
+    }
+
+    private RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> seekerForLabel( int labelId ) throws IOException
+    {
+        LabelScanKey from = new LabelScanKey().set( labelId, 0 );
+        LabelScanKey to = new LabelScanKey().set( labelId, Long.MAX_VALUE );
+        return index.seek( from, to );
+    }
+
+    private void ensureOpenCursorsClosed() throws IOException
+    {
+        RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor;
+        while ( ( cursor = openCursors.poll() ) != null )
+        {
+            cursor.close();
+        }
+    }
+
+    /**
+     * Unsupported for this implementation and will throw {@link UnsupportedOperationException}.
+     * This information can be had by asking the node store directly instead.
+     *
+     * @param nodeId node id to get labels for.
+     * @return {@link PrimitiveLongIterator} with labels for {@code nodeId}.
+     * @throws UnsupportedOperationException since not supported by this implementation.
+     */
+    @Override
+    public PrimitiveLongIterator labelsForNode( long nodeId )
+    {
+        throw new UnsupportedOperationException( "Use your db..." );
```

Comment:
This is actually not a problem because the next PR we have coming will remove that method since it's redundant anyway :)